tet.rs/src/tetris/board/my_board.rs

```diff
@@ -0,0 +1,497 @@
+        }
+
+        if self.queue.is_empty() {
+            for _ in 0..self.second_bag.len() {
+                self.queue.push_back(self.second_bag.pop_front().unwrap());
+            }
+
+            self.second_bag = self.get_random_bag();
+        }
+    }
+
+    pub fn get_queue_at(&self, index: usize) -> OptionalTypeT {
+        let max_len: usize = self.queue.len() + self.second_bag.len();
+
+        if index >= max_len {
+            return None;
+        }
+
+        let size: usize = self.queue.len();
+        if index < size {
+            Some(self.queue[index])
+        } else {
+            Some(self.second_bag[index - size])
+        }
+    }
+
+    fn pop_piece_from_queue(&mut self) -> piece_type::PieceType {
+        self.ensure_complete_queue(false);
+
+        self.queue.pop_front().unwrap()
+    }
+
+    pub fn spawn_next_piece(&mut self) -> bool {
+        let piece_type: piece_type::PieceType = self.pop_piece_from_queue();
+
+        self.has_held = false;
+
+        self.spawn_piece(piece_type)
+    }
+
+    fn spawn_piece(&mut self, piece_type: piece_type::PieceType) -> bool {
+        let spawn_position: piece_position::PiecePosition =
+            helper::get_piece_spawn_position(piece_type);
+
+        match self.put_piece_at(piece_type, spawn_position) {
+            true => {
+                self.active_piece_type = piece_type;
+                self.active_piece_position = spawn_position;
+                true
+            }
+            false => false,
+        }
+    }
+
+    fn put_piece_at(
+        &mut self,
+        piece_type: piece_type::PieceType,
+        piece_position: piece_position::PiecePosition,
+    ) -> bool {
+        let px: piece_position::PositionT = piece_position.get_x();
+        let py: piece_position::PositionT = piece_position.get_y();
+
+        let piece_matrix: piece_matrix::MatrixT =
+            piece_matrix::get_matrix_for(piece_type, piece_position.get_rotation());
+
+        let piece_size: usize = piece_matrix::get_piece_size(piece_type);
+
+        let save: BoardT = self.board.clone();
+
+        for y in 0..piece_size {
+            for x in 0..piece_size {
+                if !piece_matrix::test_bit(
+                    piece_matrix,
+                    piece_type,
+                    x as piece_position::PositionT,
+                    y as piece_position::PositionT,
+                ) {
+                    continue;
+                }
+
+                let x: piece_position::PositionT = x as piece_position::PositionT;
+                let y: piece_position::PositionT = y as piece_position::PositionT;
+
+                // FIXME: properly check both validity of position and
+                // board state at this position
+                if x + px >= BOARD_WIDTH
+                    || y + py >= BOARD_HEIGHT * 2
+                    || self.get_cell(x + px, y + py).is_none()
+                    || self.get_cell(x + px, y + py).unwrap()
```

Comment:
```suggestion
                if self.get_cell(x + px, y + py) != Some(false)
```

Reply:
I may have forgotten to clean old cold on this one...